cpu/stm32/periph/adc_l1.c

```diff
@@ -157,12 +155,18 @@ int32_t adc_sample(adc_t line, adc_res_t res)
     /* lock and power on the ADC device  */
     prep();
 
-    /* set resolution, conversion channel and single read */
+    /* mask and set resolution, conversion channel and single read */
+    ADC1->CR1 &= ~ADC_CR1_RES_Msk;
     ADC1->CR1 |= res & ADC_CR1_RES;
```

Comment:
this hould result in 1 less read write
```suggestion
    /* mask and set resolution, conversion channel and single read */
    ADC1->CR1 = (res & ADC_CR1_RES) | ( ADC1->CR1 & ~ADC_CR1_RES_Msk);
```

Reply:
I assumed that the compiler would optimize that anyways, but I'll add your suggestion.

My personal preference is to put the masking in front of everything, because then you can concatenate register modifications after that.

```suggestion
    /* mask and set resolution, conversion channel and single read */
    ADC1->CR1 = (ADC1->CR1 & ~ADC_CR1_RES_Msk) | (res & ADC_CR1_RES);
```

Reply:
I agree that readability would be better with separate statements. The issue is that when `volatile` variables are involved, every read and write operation is performed as specified.

```C
#include <stdint.h>

extern volatile uint32_t foo_reg;

void update_field_in_foo(uint8_t new_field_value)
{
    /* clear out old field value */
    foo_reg &= ~(0xff);
    /* write new field value */
    foo_reg |= new_field_value;
}

void update_field_in_foo_fast(uint8_t new_field_value)
{
    /* avoid multiple writes by using local variable for modification */
    uint32_t tmp = foo_reg;

    /* clear out old field value */
    tmp &= ~(0xff);
    /* update field new field value */
    tmp |= new_field_value;

    /* write back updated register */
    foo_reg = tmp;
}
```

Compiles to (with `arm-none-eabi-gcc` in 14.2.0 and `-Os`)

```asm
update_field_in_foo(unsigned char):
        ldr     r3, .L2
        ldr     r2, [r3]
        bic     r2, r2, #255
        str     r2, [r3]
        ldr     r2, [r3]
        orr     r0, r0, r2
        str     r0, [r3]
        bx      lr
.L2:
        .word   foo_reg
update_field_in_foo_fast(unsigned char):
        ldr     r2, .L5
        ldr     r3, [r2]
        bic     r3, r3, #255
        orr     r0, r0, r3
        str     r0, [r2]
        bx      lr
.L5:
        .word   foo_reg
```

What is particularly counter-intuitive to see (well, for me at least) is:

```asm
        str     r2, [r3]
        ldr     r2, [r3]
```

Where the value just written is read back. But that is required due to `volatile`, as the compiler must assume the value of `foo_reg` is updated externally (e.g. by hardware, because it is something like a timer counter or an SPI input register or whatever).

So: You can go for more explicit code (which I personally also consider more readable), if you first pass the value through a local variable that is not `volatile` and rely on the optimizer to work as expected for every operation on the local variable. Only the initial load from the memory reg to the local variable and the final write back are `volatile` memory operations and will not be optimized.